### PR TITLE
feat: experimental `defineRouteMeta`

### PR DESCRIPTION
**Pull request by eriksLapins**
### ❓ Type of change

- [x] 📖 Documentation (updates to the documentation, readme, or JSdoc annotations)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

I added the option to change the default provided schema directly in the server route file with defineRouteMeta, which can take in any key: value pair, with one defined in this PR - openAPI operations object, which will amend the default schema with types inferred from OperationsObject in openapi-typescript

This solves the issue that we can get better looking openapi documentation with better support

Possibly resolves the problem shown in discussion #2056 

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by eriksLapins** [comment]
Also, I wanted to add a test for the defineOpenAPISchema function, but it errors out with
`Failed to load url #internal/nitro/virtual/server-handlers` if I load in the function as below

or `Failed to load url #internal/nitro` if I load the added function from #internal/nitro/routes/openapi

My idea was to add an additional preset with something like:
```
import { describe, expect, it } from "vitest";
import { defineOpenAPISchema } from "../../src/runtime/routes/openapi";
import { useStorage } from "#internal/nitro";


describe("nitro:openapi:schema-definition", () => {
    it("stores the result in useStorage", async () => {
        defineOpenAPISchema({
            routeBase: '/api/test',
            method: 'get',
            schema: {
                tags: ["test"]
            }
        })

        const storage = await useStorage("/api/test").getItem('openapi-get');

        expect(storage).toBe({
            routeBase: '/api/test',
            method: 'get',
            schema: {
                tags: ["test"]
            }
        })
    })
})
```

File: examples/api-routes/api/test.get.ts [hunk 1]
```diff
@@ -1 +1,14 @@
+import { defineOpenAPISchema } from "#internal/nitro/routes/openapi";
+
 export default defineEventHandler(() => "Test get handler");
+
+defineOpenAPISchema({
```
Comment (pi0) [inline examples/api-routes/api/test.get.ts:5]: We need to generalize util to `defineRouteMeta({ openAPI: { ... } }`)

File: examples/api-routes/api/test.post.ts [hunk 2]
```diff
@@ -1 +1,28 @@
+import { defineOpenAPISchema } from "#internal/nitro/routes/openapi";
+
 export default defineEventHandler(() => "Test post handler");
+
+defineOpenAPISchema({
+    method: 'post',
+    routeBase: '/api/test',
```
Comment (pi0) [inline examples/api-routes/api/test.post.ts:7]: Some props such as method, and url, can be auto-generated already without user needing them

File: src/runtime/routes/openapi.ts [hunk 3]
```diff
@@ -96,3 +115,9 @@ function defaultTags(route: string) {
 
   return tags;
 }
+
+export function defineOpenAPISchema(schema: NitroOpenapiSchema) {
+  useStorage(schema.routeBase).setItem(`openapi-${schema.method}`, schema);
```
Comment (pi0) [inline src/runtime/routes/openapi.ts:120]: This is a really interesting approach however considering overhead and limitations (dev<>build communication) I think we have to use static code extraction for route meta data.

Comment (eriksLapins) [reply src/runtime/routes/openapi.ts]: I changed the approach to parsing it at build time, please check if this version seems to be on the right path at least

Comment (eriksLapins) [reply src/runtime/routes/openapi.ts]: @pi0 I redid the parse to build time, could you please check if this is better? Got some inspiration from [(Nuxt's extract route rules)](https://github.com/nuxt/nuxt/blob/95a521376614d9cf605ca16fefe80d6c216bd906/packages/nuxt/src/pages/route-rules.ts#L25)

File: src/scan.ts [hunk 4]
```diff
@@ -122,3 +134,48 @@ async function scanDir(
     })
     .sort((a, b) => a.path.localeCompare(b.path));
 }
+
+export async function scanRouteMeta(
+  fullPath: string
+): Promise<ServerRouteMeta> {
+  const file = fs.readFileSync(fullPath, { encoding: "utf8", flag: "r" });
```
Comment (septatrix) [inline src/scan.ts:141]: Have you investigated how this performs for large projects with many routes and big files?

**Review by septatrix — COMMENTED** [review]
Definitely looks like a nice improvement. I am still mostly looking forward to automatically inferring the type (see #938) but this is a good compromise for the time being.

What I do not like is that is seems to only support object literals (based on reading the code, please correct me if I am wrong). No real evaluation for the route metadata is possible. This can and will confuse a lot of people. My first though was that this could be used with zod to generate zome of the types automatically but this does not seem feasible.

File: src/scan.ts (hunk 4 above)
Comment (pi0) [reply src/scan.ts]: (this is updated now)

**Comment by pi0** [comment]
Thanks for your initiatives on this dear @eriksLapins. I have made some refactors in the way it works to experiment alternative approach for Nitro. 

The meta is scanned using a rollup plugin now to allow HMR and uses a native parser of rollup + AST to Object transform and is only enabled with an experimental open API flag.

**Comment by Atinux** [comment]
How can we share the validation (let's say I am using zod) between the route meta and the `readValidatedBody` for example?

**Comment by pi0** [comment]
It is experimental and not for validation right now.

**Comment by rmarquet21** [comment]
@pi0 Any idea when this will be in production and not just on the "nightly" version?

**Comment by chenyuncai** [comment]
Currentlly it cannot collect nested router endpoint, can we hack the router difinition and get all defined api?

**Comment by pi0** [comment]
2.10 release no ETA. Also mind you, this API is likely to completely change into event handler object format nothing to rely on.

**Pull request merged**